src/d3d9/d3d9_mem.cpp

```diff
@@ -0,0 +1,269 @@
+    std::lock_guard<dxvk::mutex> lock(m_mutex);
+
+    m_usedMemory -= Size;
+  }
+
+  void D3D9MemoryAllocator::FreeChunk(D3D9MemoryChunk *Chunk) {
+    std::lock_guard<dxvk::mutex> lock(m_mutex);
+
+    m_allocatedMemory -= Chunk->Size();
+    m_chunks.erase(std::remove(m_chunks.begin(), m_chunks.end(), Chunk));
+    delete Chunk;
+  }
+
+  uint32_t D3D9MemoryAllocator::MappedMemory() {
+    std::lock_guard<dxvk::mutex> lock(m_mutex);
+
+    return m_mappedMemory;
+  }
+
+  uint32_t D3D9MemoryAllocator::UsedMemory() {
+    std::lock_guard<dxvk::mutex> lock(m_mutex);
+
+    return m_usedMemory;
+  }
+
+  D3D9MemoryChunk::D3D9MemoryChunk(D3D9MemoryAllocator* Allocator, uint32_t Size)
+    : m_allocator(Allocator), m_size(Size) {
+    if constexpr (!env::is32BitHostPlatform()) {
+      m_ptr = malloc(Size);
+      m_mapping = nullptr;
+      return;
+    }
+
+    m_mapping = CreateFileMappingA(INVALID_HANDLE_VALUE, nullptr, PAGE_READWRITE | SEC_COMMIT, 0, Size, nullptr);
+    m_freeRanges.push_back({ 0, Size });
+  }
+
+  D3D9MemoryChunk::~D3D9MemoryChunk() {
+    if constexpr (!env::is32BitHostPlatform()) {
+      free(m_ptr);
+      return;
+    }
+
+    if (m_ptr != nullptr) {
+      UnmapViewOfFile(m_ptr);
+    }
+    CloseHandle(m_mapping);
+  }
+
+  void D3D9MemoryChunk::IncMapCounter() {
+    if constexpr (!env::is32BitHostPlatform())
+      return;
+
+    std::lock_guard<dxvk::mutex> lock(m_mutex);
+    m_mapCounter++;
+
+    if (m_mapCounter == 1) {
+      m_allocator->NotifyMapped(m_size);
+      m_ptr = MapViewOfFile(m_mapping, FILE_MAP_ALL_ACCESS, 0, 0, m_size);
+    }
+  }
+
+  void D3D9MemoryChunk::DecMapCounter() {
+    if constexpr (!env::is32BitHostPlatform())
+      return;
+
+    std::lock_guard<dxvk::mutex> lock(m_mutex);
+    m_mapCounter--;
+
+    if (m_mapCounter == 0) {
+      m_allocator->NotifyUnmapped(m_size);
+      UnmapViewOfFile(m_ptr);
+      m_ptr = nullptr;
+    }
+  }
+
+  D3D9Memory D3D9MemoryChunk::Alloc(uint32_t Size) {
+    if constexpr (!env::is32BitHostPlatform()) {
+      // In 64bit builds we use a 1:1 mapping between chunks and memory
+      return D3D9Memory(this, 0, Size);
+    }
+
+    std::lock_guard<dxvk::mutex> lock(m_mutex);
+
+    uint32_t offset = 0;
+    uint32_t size = 0;
+
+    for (auto range = m_freeRanges.begin(); range != m_freeRanges.end(); range++) {
+      if (range->length >= Size) {
+        offset = range->offset;
+        size = Size;
+        range->offset += Size;
+        range->length -= Size;
+        if (range->length < (4 << 10)) {
+          size += range->length;
+          m_freeRanges.erase(range);
+        }
+        break;
+      }
+    }
+
+    if (size != 0)
+      return D3D9Memory(this, offset, size);
+
+    return {};
+  }
+
+  void D3D9MemoryChunk::Free(D3D9Memory *Memory) {
+    if constexpr (!env::is32BitHostPlatform()) {
+      // In 64bit builds we use a 1:1 mapping between chunks and memory
+      free(m_ptr);
+      delete this;
```

Comment:
I hate everything about this. You can't simply commit suicide, you don't even know that this was allocated with `new`.

Reply:
Imo, call `m_allocator->FreeChunk(this);` and that could call delete. But even that's still a bit eww.